integration_tests/CMakeLists.txt

```diff
@@ -279,7 +279,7 @@ RUN(NAME test_list_section   LABELS cpython llvm c)
 RUN(NAME test_tuple_01       LABELS cpython llvm c)
 RUN(NAME test_tuple_02       LABELS cpython llvm c)
 RUN(NAME test_tuple_03       LABELS cpython llvm)
-RUN(NAME test_dict_01        LABELS cpython llvm)
+RUN(NAME test_dict_01        LABELS cpython llvm c)
```

Comment:
How does the C code looks like for this?

Reply:
Here it is:

<details>
<summary> C code </summary>
<p>

```c
#include <complex.h>
#include <inttypes.h>

#include <stdlib.h>
#include <stdbool.h>
#include <stdio.h>
#include <string.h>
#include <lfortran_intrinsics.h>

#define ASSERT(cond)                                                           \
    {                                                                          \
        if (!(cond)) {                                                         \
            printf("%s%s", "ASSERT failed: ", __FILE__);                       \
            printf("%s%s", "\nfunction ", __func__);                           \
            printf("%s%d%s", "(), line number ", __LINE__, " at \n");          \
            printf("%s%s", #cond, "\n");                                       \
            exit(1);                                                           \
        }                                                                      \
    }
#define ASSERT_MSG(cond, msg)                                                  \
    {                                                                          \
        if (!(cond)) {                                                         \
            printf("%s%s", "ASSERT failed: ", __FILE__);                       \
            printf("%s%s", "\nfunction ", __func__);                           \
            printf("%s%d%s", "(), line number ", __LINE__, " at \n");          \
            printf("%s%s", #cond, "\n");                                       \
            printf("%s", "ERROR MESSAGE:\n");                                  \
            printf("%s%s", msg, "\n");                                         \
            exit(1);                                                           \
        }                                                                      \
    }


struct dimension_descriptor
{
    int32_t lower_bound, length;
};

struct dict_i32_r64_ {
    int32_t *key;
    double *value;
    int capacity;
    bool *present;
};

inline void dict_init_dict_i32_r64_(struct dict_i32_r64_* x, int32_t capacity);
inline void dict_resize_dict_i32_r64_(struct dict_i32_r64_* x);
inline void dict_insert_dict_i32_r64_(struct dict_i32_r64_* x, int32_t k,double v);
inline double dict_get_item_dict_i32_r64_(struct dict_i32_r64_* x, int32_t k);
inline int32_t dict_len_dict_i32_r64_(struct dict_i32_r64_* x);
inline void dict_deepcopy_dict_i32_r64_(struct dict_i32_r64_* src, struct dict_i32_r64_* dest);


// Implementations
double __lpython_overloaded_0__abs(double x)
{
    double _lpython_return_variable;
    double result;
    if (x >=   0.00000000000000000e+00) {
        result = x;
    } else {
        result = -x;
    }
    _lpython_return_variable = result;
    return _lpython_return_variable;
}

float _lfortran_caimag(float complex x);

double _lfortran_zaimag(double complex x);

void test_dict()
{
    int32_t i;
    struct dict_i32_r64_ rollnumber2cpi;
    int32_t size = 1000;
    struct dict_i32_r64_ constname0;
    dict_init_dict_i32_r64_(&constname0, 0 + 1);
    dict_deepcopy_dict_i32_r64_(&constname0, &rollnumber2cpi);
    size = 1000;
    dict_insert_dict_i32_r64_(&rollnumber2cpi, 0,   1.10000000000000009e+00);
    for (i=1000; i<=1000 + size - 1; i++) {
        dict_insert_dict_i32_r64_(&rollnumber2cpi, i, (double)(i)/  1.00000000000000000e+02 +   5.00000000000000000e+00);
    }
    for (i=1000 + size - 1; i>=1001 + 1; i--) {
        {
            ASSERT(__lpython_overloaded_0__abs(dict_get_item_dict_i32_r64_(&rollnumber2cpi, i) - (double)(i)/  1.00000000000000000e+02 -   5.00000000000000000e+00) <=   9.99999999999999980e-13);
        }
    }
    ASSERT(__lpython_overloaded_0__abs(dict_get_item_dict_i32_r64_(&rollnumber2cpi, 0) -   1.10000000000000009e+00) <=   9.99999999999999980e-13);
    ASSERT(dict_len_dict_i32_r64_(&rollnumber2cpi) == 1001);
}

void _lpython_main_program()
{
    test_dict();
}

int main(int argc, char* argv[])
{
    _lpython_main_program();
    return 0;
}

void dict_init_dict_i32_r64_(struct dict_i32_r64_* x, int32_t capacity) {
    x->capacity = capacity;
    x->key = (int32_t*) malloc(capacity * sizeof(int32_t));
    x->value = (double*) malloc(capacity * sizeof(double));
    x->present = (bool*) malloc(capacity * sizeof(bool));
    memset(x->present, false,capacity * sizeof(bool));
}

void dict_resize_dict_i32_r64_(struct dict_i32_r64_* x) {
    int32_t *tmp_key = (int32_t *) malloc(x->capacity * sizeof(int32_t));
    memcpy(tmp_key, x->key, x->capacity * sizeof(int32_t));
    double *tmp_val = (double *) malloc(x->capacity * sizeof(double));
    memcpy(tmp_val, x->value, x->capacity * sizeof(double));
    bool *tmp_p = (bool *) malloc(x->capacity * sizeof(bool));
    memcpy(tmp_p, x->present, x->capacity * sizeof(bool));
    x->capacity = 2*x->capacity+1;
    free(x->key); free(x->value); free(x->present);
    x->key = (int32_t*) malloc(x->capacity * sizeof(int32_t));
    x->value = (double*) malloc(x->capacity * sizeof(double));
    x->present = (bool*) malloc(x->capacity * sizeof(bool));
    memset(x->present, false,x->capacity * sizeof(bool));
    for(size_t i=0; i<x->capacity/2; i++) {
        if(tmp_p[i]) {
            int j=tmp_key[i]%x->capacity;
            while(x->present[j]) j=(j+1)%x->capacity;
            x->key[j] = tmp_key[i]; x->value[j] = tmp_val[i]; x->present[j] = true;
        }
    }
    free(tmp_key); free(tmp_val); free(tmp_p);
}

void dict_insert_dict_i32_r64_(struct dict_i32_r64_* x, int32_t k,double v) {
    int j=k%x->capacity; int c = 0;
    while(c < x->capacity && x->present[j]) j=(j+1)%x->capacity, c++;
    if (c == x->capacity) {
        dict_resize_dict_i32_r64_(x);
        j=k%x->capacity;
        while(x->present[j]) j=(j+1)%x->capacity;
    }
    x->key[j] = k; x->value[j] = v; x->present[j] = true;
}

double dict_get_item_dict_i32_r64_(struct dict_i32_r64_* x, int32_t k) {
    int j=k%x->capacity;
    while(x->present[j] && !(x->key[j] == k)) j=(j+1)%x->capacity;
    if (x->present[j] && x->key[j] == k) return x->value[j];
    printf("Key not found");
    exit(1);
}

int32_t dict_len_dict_i32_r64_(struct dict_i32_r64_* x) {
    int32_t len = 0;
    for(int i=0; i<x->capacity; i++) len += (int)x->present[i];
    return len;
}

void dict_deepcopy_dict_i32_r64_(struct dict_i32_r64_* src, struct dict_i32_r64_* dest) {
    dest->capacity = src->capacity;
    dest->key = (int32_t*) malloc(dest->capacity * sizeof(int32_t));
    dest->value = (double*) malloc(dest->capacity * sizeof(double));
    dest->present = (bool*) malloc(dest->capacity * sizeof(bool));
    memcpy(dest->key, src->key, src->capacity * sizeof(int32_t));
    memcpy(dest->value, src->value, src->capacity * sizeof(double));
    memcpy(dest->present, src->present, src->capacity * sizeof(bool));
}

```

</p>
</details>

Reply:
LGTM.

Reply:
Would be good to see it benchmarked. :-). Let's see how this one performs.

Reply:
These are the results on my machine (Apple M1 2020):

<details>
<summary> b.c (Generated from lpython - commented lfortran_intrinsics.h) </summary>
<p>

```c
#include <complex.h>
#include <inttypes.h>

#include <stdlib.h>
#include <stdbool.h>
#include <stdio.h>
#include <string.h>
// #include < lfortran_intrinsics.h>

#define ASSERT(cond)                                                           \
    {                                                                          \
        if (!(cond)) {                                                         \
            printf("%s%s", "ASSERT failed: ", __FILE__);                       \
            printf("%s%s", "\nfunction ", __func__);                           \
            printf("%s%d%s", "(), line number ", __LINE__, " at \n");          \
            printf("%s%s", #cond, "\n");                                       \
            exit(1);                                                           \
        }                                                                      \
    }
#define ASSERT_MSG(cond, msg)                                                  \
    {                                                                          \
        if (!(cond)) {                                                         \
            printf("%s%s", "ASSERT failed: ", __FILE__);                       \
            printf("%s%s", "\nfunction ", __func__);                           \
            printf("%s%d%s", "(), line number ", __LINE__, " at \n");          \
            printf("%s%s", #cond, "\n");                                       \
            printf("%s", "ERROR MESSAGE:\n");                                  \
            printf("%s%s", msg, "\n");                                         \
            exit(1);                                                           \
        }                                                                      \
    }


struct dimension_descriptor
{
    int32_t lower_bound, length;
};

struct dict_i32_r64_ {
    int32_t *key;
    double *value;
    int capacity;
    bool *present;
};

inline void dict_init_dict_i32_r64_(struct dict_i32_r64_* x, int32_t capacity);
inline void dict_resize_dict_i32_r64_(struct dict_i32_r64_* x);
inline void dict_insert_dict_i32_r64_(struct dict_i32_r64_* x, int32_t k,double v);
inline double dict_get_item_dict_i32_r64_(struct dict_i32_r64_* x, int32_t k);
inline int32_t dict_len_dict_i32_r64_(struct dict_i32_r64_* x);
inline void dict_deepcopy_dict_i32_r64_(struct dict_i32_r64_* src, struct dict_i32_r64_* dest);


// Implementations
double __lpython_overloaded_0__abs(double x)
{
    double _lpython_return_variable;
    double result;
    if (x >=   0.00000000000000000e+00) {
        result = x;
    } else {
        result = -x;
    }
    _lpython_return_variable = result;
    return _lpython_return_variable;
}

float _lfortran_caimag(float complex x);

double _lfortran_zaimag(double complex x);

void test_dict()
{
    int32_t i;
    struct dict_i32_r64_ rollnumber2cpi;
    int32_t size = 1000;
    struct dict_i32_r64_ constname0;
    dict_init_dict_i32_r64_(&constname0, 0 + 1);
    dict_deepcopy_dict_i32_r64_(&constname0, &rollnumber2cpi);
    size = 1000;
    dict_insert_dict_i32_r64_(&rollnumber2cpi, 0,   1.10000000000000009e+00);
    for (i=1000; i<=1000 + size - 1; i++) {
        dict_insert_dict_i32_r64_(&rollnumber2cpi, i, (double)(i)/  1.00000000000000000e+02 +   5.00000000000000000e+00);
    }
    for (i=1000 + size - 1; i>=1001 + 1; i--) {
        {
            ASSERT(__lpython_overloaded_0__abs(dict_get_item_dict_i32_r64_(&rollnumber2cpi, i) - (double)(i)/  1.00000000000000000e+02 -   5.00000000000000000e+00) <=   9.99999999999999980e-13);
        }
    }
    ASSERT(__lpython_overloaded_0__abs(dict_get_item_dict_i32_r64_(&rollnumber2cpi, 0) -   1.10000000000000009e+00) <=   9.99999999999999980e-13);
    ASSERT(dict_len_dict_i32_r64_(&rollnumber2cpi) == 1001);
}

void _lpython_main_program()
{
    test_dict();
}

int main(int argc, char* argv[])
{
    _lpython_main_program();
    return 0;
}

void dict_init_dict_i32_r64_(struct dict_i32_r64_* x, int32_t capacity) {
    x->capacity = capacity;
    x->key = (int32_t*) malloc(capacity * sizeof(int32_t));
    x->value = (double*) malloc(capacity * sizeof(double));
    x->present = (bool*) malloc(capacity * sizeof(bool));
    memset(x->present, false,capacity * sizeof(bool));
}

void dict_resize_dict_i32_r64_(struct dict_i32_r64_* x) {
    int32_t *tmp_key = (int32_t *) malloc(x->capacity * sizeof(int32_t));
    memcpy(tmp_key, x->key, x->capacity * sizeof(int32_t));
    double *tmp_val = (double *) malloc(x->capacity * sizeof(double));
    memcpy(tmp_val, x->value, x->capacity * sizeof(double));
    bool *tmp_p = (bool *) malloc(x->capacity * sizeof(bool));
    memcpy(tmp_p, x->present, x->capacity * sizeof(bool));
    x->capacity = 2*x->capacity+1;
    free(x->key); free(x->value); free(x->present);
    x->key = (int32_t*) malloc(x->capacity * sizeof(int32_t));
    x->value = (double*) malloc(x->capacity * sizeof(double));
    x->present = (bool*) malloc(x->capacity * sizeof(bool));
    memset(x->present, false,x->capacity * sizeof(bool));
    for(size_t i=0; i<x->capacity/2; i++) {
        if(tmp_p[i]) {
            int j=tmp_key[i]%x->capacity;
            while(x->present[j]) j=(j+1)%x->capacity;
            x->key[j] = tmp_key[i]; x->value[j] = tmp_val[i]; x->present[j] = true;
        }
    }
    free(tmp_key); free(tmp_val); free(tmp_p);
}

void dict_insert_dict_i32_r64_(struct dict_i32_r64_* x, int32_t k,double v) {
    int j=k%x->capacity; int c = 0;
    while(c < x->capacity && x->present[j]) j=(j+1)%x->capacity, c++;
    if (c == x->capacity) {
        dict_resize_dict_i32_r64_(x);
        j=k%x->capacity;
        while(x->present[j]) j=(j+1)%x->capacity;
    }
    x->key[j] = k; x->value[j] = v; x->present[j] = true;
}

double dict_get_item_dict_i32_r64_(struct dict_i32_r64_* x, int32_t k) {
    int j=k%x->capacity;
    while(x->present[j] && !(x->key[j] == k)) j=(j+1)%x->capacity;
    if (x->present[j] && x->key[j] == k) return x->value[j];
    printf("Key not found");
    exit(1);
}

int32_t dict_len_dict_i32_r64_(struct dict_i32_r64_* x) {
    int32_t len = 0;
    for(int i=0; i<x->capacity; i++) len += (int)x->present[i];
    return len;
}

void dict_deepcopy_dict_i32_r64_(struct dict_i32_r64_* src, struct dict_i32_r64_* dest) {
    dest->capacity = src->capacity;
    dest->key = (int32_t*) malloc(dest->capacity * sizeof(int32_t));
    dest->value = (double*) malloc(dest->capacity * sizeof(double));
    dest->present = (bool*) malloc(dest->capacity * sizeof(bool));
    memcpy(dest->key, src->key, src->capacity * sizeof(int32_t));
    memcpy(dest->value, src->value, src->capacity * sizeof(double));
    memcpy(dest->present, src->present, src->capacity * sizeof(bool));
}
```

</p>
</details>


Clang
```
% gcc -v
Apple clang version 14.0.0 (clang-1400.0.29.202)
Target: arm64-apple-darwin22.1.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
% gcc b.c -o b.o
% time ./b.o
./b.o  0.00s user 0.00s system 0% cpu 0.623 total
% time ./b.o
./b.o  0.00s user 0.00s system 62% cpu 0.007 total
% time ./b.o
./b.o  0.00s user 0.00s system 63% cpu 0.008 total
% time ./b.o
./b.o  0.00s user 0.00s system 62% cpu 0.007 total
```

GNU
```
% gcc-12 -v
Using built-in specs.
COLLECT_GCC=gcc-12
COLLECT_LTO_WRAPPER=/opt/homebrew/Cellar/gcc/12.1.0/bin/../libexec/gcc/aarch64-apple-darwin21/12/lto-wrapper
Target: aarch64-apple-darwin21
Configured with: ../configure --prefix=/opt/homebrew/opt/gcc --libdir=/opt/homebrew/opt/gcc/lib/gcc/current --disable-nls --enable-checking=release --with-gcc-major-version-only --enable-languages=c,c++,objc,obj-c++,fortran --program-suffix=-12 --with-gmp=/opt/homebrew/opt/gmp --with-mpfr=/opt/homebrew/opt/mpfr --with-mpc=/opt/homebrew/opt/libmpc --with-isl=/opt/homebrew/opt/isl --with-zstd=/opt/homebrew/opt/zstd --with-pkgversion='Homebrew GCC 12.1.0' --with-bugurl=https://github.com/Homebrew/homebrew-core/issues --with-system-zlib --build=aarch64-apple-darwin21 --with-sysroot=/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 12.1.0 (Homebrew GCC 12.1.0) 
% gcc-12 b.c -o b.o
% time ./b.o
./b.o  0.00s user 0.00s system 1% cpu 0.308 total
% time ./b.o
./b.o  0.00s user 0.00s system 62% cpu 0.007 total
% time ./b.o
./b.o  0.00s user 0.00s system 59% cpu 0.007 total
```

LPython
```
% lpython integration_tests/test_dict_01.py -o a.o
% time ./a.o
./a.o  0.01s user 0.01s system 0% cpu 1.565 total
% time ./a.o
./a.o  0.01s user 0.01s system 63% cpu 0.018 total
% time ./a.o
./a.o  0.01s user 0.00s system 62% cpu 0.016 total
% time ./a.o
./a.o  0.01s user 0.00s system 60% cpu 0.017 total
```

Reply:
You have to do `lpython --fast`.

Also, you need to add your terminal as an exception in security policy in settings, so that it does not check the hash of your new executable online. That's why it sometimes takes 1.5s to run for the first time for you. After the fix it will run immediately.